부하테스트 자동화, 인덱스를 이용해 API 성능 개선

donggi-lee-bit opened the pull request:
### ❗️ 이슈 번호

Closes #52 
Closes #53 


<br><br>

### 📝 구현 내용

- `AWS EC2` 에 nGrinder Controller, agent, 테스트 서버 구축
- `GitHub Actions` 를 이용해 부하테스트 자동화 워크플로우 실행
  - `aws cli` 를 이용해 EC2 실행
  - nGrinder Rest API로 부하테스트 실행 요청
  - 부하테스트 결과 디스코드로 전송
  - `aws cli` 로 EC2 stop (자원 정리)

[부하테스트 자동화 정리](https://victorious-lint-a8e.notion.site/GitHub-Actions-aws-cli-f34561111a544f62a0dfde21fe3ba9ac)

- 인덱스를 이용해 API 성능 개선
  - 댓글(comment) 테이블의 게시글 id(comma_id) 컬럼에 인덱스 추가
    - MySQL 실행 계획으로 테이블을 풀스캔하고 있는지, 불필요하게 데이터에 얼마나 접근하는지 확인
    - 인덱스를 걸어 부하테스트 시 TPS 값이 `0.2 -> 18.1` 로 상승된 것 확인
  - 회고 게시글(Comma) 테이블에 created_at 컬럼에 인덱스 추가
    - 특정 날짜에 작성된 회고 글 조회 API 부하테스트 시 TPS 값 `4.8 -> 158.5` 로 상승된 것 확인

이전 PR에서 로컬 환경에서 부하테스트를 해보면서 배포 환경에서도 부하테스트를 해보고 싶다는 욕구가 있었습니다. 그 욕구가 부하테스트 자동화까지 가게 되었는데, 그 과정에서 많은 경험을 하였습니다.
- AWS RDS, EC2 자원 관리
- AWS CLI를 사용할 때 `aws configure` 값 설정
- 쉘 스크립트로 EC2의 public IP를 수집하여 nGrinder controller, agent, 테스트 서버를 실행하기 
- nGrinder Rest API 로 통신하여 부하테스트 실행, 결과 조회

등 기존에 GitHub Actions 로 CI/CD 외의 작업을 자동화해보니 신기하고도 재밌는 경험이었습니다.

그리고 가장 중요한 성능 개선을 해보았는데 인덱스를 걸면서도 여러 시행 착오가 있었고 아직도 알아가는 중입니다. [업무에 바로 쓰는 SQL 튜닝을 읽고 프로젝트 API 쿼리 튜닝 시행 착오 기록](https://victorious-lint-a8e.notion.site/SQL-Comma-ea373c0a0e9e4b2bbe6ef8643a08f0c6)

기한을 지키는 것도 중요한데 매번 생각한 기한보다 늦어져 아쉽네요 😅
이번에도 잘 부탁드리겠습니다 🥹

<img width="660" alt="image" src="https://user-images.githubusercontent.com/73376468/233606052-c936a420-9613-46d6-a002-3020e5e2601e.png">

\+ 추가
날짜 기반 회고 게시글 검색 조회 API 에 커버링 인덱스를 적용하여 성능을 조금 더 개선하였습니다. 
전 후 TPS 158 -> 558

<img width="1232" alt="image" src="https://user-images.githubusercontent.com/73376468/234837390-330edb86-5993-4944-8159-861e1a3c7622.png">

donggi-lee-bit committed:
refactor: 성공 실패 디스코드 알림 색상 변경

- 성공 시 초록색, 실패 시 빨간색으로 변경

donggi-lee-bit committed:
refactor: 전체 게시글 조회 시 페이지네이션 사이즈를 변수로 관리하도록 변경

- 게시글 조회 시 최대 요청 사이즈는 100개로 지정

donggi-lee-bit committed:
refactor: 전체 댓글 조회 시 페이지네이션 사이즈를 변수로 관리하도록 변경

- 댓글 전체 조회 시 최대 요청 사이즈는 10개로 지정

donggi-lee-bit committed:
refactor: 댓글 엔티티에서 회고 게시글 id 인덱싱

- 댓글 엔티티에 회고 게시글 id를 인덱스 걸어 조회 속도 향상 기대
- 실제 sql 쿼리 실행시 1.61 -> 0.01 초로 향상

donggi-lee-bit committed:
perf: 회고 게시글 엔티티에 index 어노테이션으로 인덱스 추가

- comma 테이블의 created_at 컬럼에 인덱스를 걸어주었음
- 실행 계획, 부하테스트 TPS 상승으로 특정 날짜에 작성된 회고 게시글 조회 API 성능 향상 확인

donggi-lee-bit committed:
perf: 커버링 인덱스를 이용해 회고 게시글 작성된 날짜 기준 검색 조회 API 성능 개선

- 커버링 인덱스로 인덱스 검색을 먼저 처리하고, 걸러진 데이터에 대해서만 데이터 블록에 접근하여 성능의 이점을 얻게 됨

ku-kim reviewed (changes requested):
페이징의 커서가 추가되어야 할 거 같아요~

donggi-lee-bit committed:
refactor: 회고 게시글 페이징 메타데이터 추가

- 페이징 데이터와 회고 게시글 데이터를 응답하는 CommaPaginatedResponse 클래스 추가
- 페이징 메타데이터가 추가됨으로 API 명세 최신화를 위해 SpringRestDocs Controller 테스트 수정

donggi-lee-bit committed:
perf: 특정 유저가 작성한 게시글 조회 API 커버링 인덱스 적용

donggi-lee-bit committed:
perf: 특정 유저, 특정 날짜에 작성된 게시글 조회 API 커버링 인덱스 적용

donggi-lee-bit committed:
refactor: 게시글 검색 조회 API count 쿼리 추가

- 페이지 번호를 매기기 위한 count 쿼리 추가

donggi-lee-bit committed:
refactor: 동적 쿼리를 통해 검색 조건 게시글 반환하고,

- QueryDsl 의 BooleanExpression 을 이용하여 date, username 값이 있거나 null 인 것에 따라 쿼리를 동적으로 실행하도록 변경
- PageableDefault 어노테이션으로 페이지 기본 값 설정
  - 기본 값 : page = 0, size = 10

donggi-lee-bit committed:
refactor: 게시글, 댓글 max 페이지 사이즈 예외처리 방법 변경

- max 사이즈보다 초과된 사이즈로 요청 시 예외를 반환하도록 했었지만 max 사이즈로 바꾸어서 데이터 반환할 수 있도록 변경
- max size : 게시글은 100, 댓글은 10

donggi-lee-bit committed:
refactor: 댓글 검색 페이지 max 사이즈 방어 로직 및 로직 변경에 따른 테스트 코드 변경

donggi-lee-bit commented:
@ku-kim 

안녕하세요 리뷰어님~ **페이지 메타데이터 추가**, **회고 게시글 검색 동적 쿼리로 동작하도록 변경**, **페이지 사이즈 예외처리가 아닌 방어 로직으로 동작하도록 변경** 작업 추가하였습니다.
페이지 메타데이터로는 현재 페이지, 반환되는 데이터 개수, 전체 페이지(마지막 페이지 위치로 사용)를 추가해주었습니다. 
또한 날짜, 작성자 이름, 날짜와 작성자 이름 세 가지의 게시글 검색 조건을 동적 쿼리로 동작하도록 개선하였습니다.
기존에 요청한 페이지 사이즈가 일정 크기 이상(게시글 : 100개, 댓글 : 10개)일 경우 예외를 반환하도록 하였는데 페이지 사이즈로 예외를 반환하는 로직의 흐름이 이상하다고 판단하여 정해놓은(게시글 : 100개, 댓글 : 10개) 최대 크기로 데이터를 반환하도록 변경하였습니다.

donggi-lee-bit committed:
chore: DB 실행을 위한 Dockerfile 추가

- 모니터링 프로그램을 실행하는 도커 컴포즈 파일에 DB 실행 명령어 추가

donggi-lee-bit committed:
refactor: 프로메테우스 target 변경

- 네트워크 문제로 공유기 ip 로 매번 수정해주어야했던 호스트를 docker.for.mac.localhost 로 해결

donggi-lee-bit committed:
refactor: 도커 컴포즈가 위치한 디렉토리명 변경

donggi-lee-bit committed:
docs: 애플리케이션 실행 가이드 리드미 추가

ku-kim reviewed (approved):
대면 리뷰 check,
Pageable global 옵션 설정과 상수 처리, Sort 유효성 검증 정도 추가하시면 좋을 거 같습니다~

donggi-lee-bit committed:
refactor: 페이지 방어로직 서비스 단에서 ArgumentResolver 에서 처리하도록 변경

- 서비스 계층에서는 비즈니스 로직에 집중하고 ArgumentResolver 에서 예외처리 하도록 변경
- @ValidPage 어노테이션 생성
- 페이지 사이즈가 100 보다 클 시 최대 페이지 사이즈인 100으로 사이즈 강제 변경
- 페이지 값 없이 요청 시 기본 페이지 값으로 1로 처리
- 페이지 사이즈 값 없이 요청 시 기본 페이지 사이즈인 10으로 처리
- CustomPageableArgumentResolver 테스트 추가

donggi-lee-bit committed:
feat: 인터셉터에서 페이지 정렬 예외 처리

- 정렬 작업은 서버에 부하가 많이 생기는 작업임으로 앞단에서 임의로 정렬할 수 없도록 인터셉터에서 예외를 발생시키도록 처리

donggi-lee-bit committed:
refactor: 페이지 번호 0부터 시작하도록 one-indexed-parameter 옵션 삭제

donggi-lee-bit commented:
@ku-kim 

피드백 받은 작업 추가하여 Merge 하겠습니다!
- 데이터베이스 실행 파일 추가
- 페이지 전역 설정
- 서비스 로직에서 페이지 방어 로직을 ArgumentResolver 에서 처리하도록 변경
- 페이지 정렬 방어 로직 Interceptor 에서 처리하도록 개발


다음 PR 에 해당 작업 내용도 포함하도록 하겠습니다 😆
